### DISCUSSION TITLE
Add moderation request for OpenAI and TikTokenUtils for count tokens

### DISCUSSION THREAD

**Discussion by kkaletka** [spring-projects/spring-ai#509] [Ideas]
Hello,

As I've been exploring the capabilities of your Spring AI library, I've encountered a few challenges. Notably, unlike the implementation seen in Theo Kanning's openai library, there are no directly available methods for engaging with moderation services—where dtos are crafted for the response alongside a choice of moderation models. Are there any intentions to incorporate such functionality in the future? Currently, to achieve this, I find myself manually crafting requests with RestTemplate and constructing response classes.

Additionally, I'm curious if there are any plans to introduce something akin to TikTokenUtils, as found in theokanning.openai's com.theokanning.openai.utils.TikTokensUtil. This utility class is exceptionally handy for calculating the expected token consumption of a prompt before making the request.

Thank you for considering these inquiries.

**Comment by dafriz** [comment]
Replying to your additional question, you can use the `JTokkitTokenCountEstimator` in spring-ai-core to estimate the expected token consumption. You should get similar results as Theo Kanning's utility as they both use the JTokkit library to perform the calculation.

**Comment by dafriz** [comment]
In regard to your original quesiton, check out the `OpenAiModerationModel` and associated classes which were added to Spring AI after you raised this. See documentation here - https://docs.spring.io/spring-ai/reference/api/moderation/openai-moderation.html
